dev/src/bulk-writer.ts

```diff
@@ -336,24 +315,76 @@ export class BulkWriter {
   /**
    * The maximum number of writes that can be in a single batch.
    */
-  private maxBatchSize = MAX_BATCH_SIZE;
+  private _maxBatchSize = MAX_BATCH_SIZE;
 
   /**
    * A queue of batches to be written.
    */
-  private batchQueue: BulkCommitBatch[] = [];
+  private _batchQueue: BulkCommitBatch[] = [];
+
+  /**
+   * A queue of batches containing operations that need to be retried.
+   */
+  private _retryBatchQueue: BulkCommitBatch[] = [];
 
   /**
-   * Whether this BulkWriter instance is closed. Once closed, it cannot be
-   * opened again.
+   * A list of promises that represent sent batches. Each promise is resolved
+   * when the batch's response is received. This includes batches from both the
+   * batchQueue and retryBatchQueue.
    */
-  private closed = false;
+  private _pendingBatches: Set<Promise<void>> = new Set();
+
+  /**
+   * A list of promises that represent pending BulkWriter operations. Each
+   * promise is resolved when the BulkWriter operation resolves. This set
+   * includes retries. Each retry's promise is added, attempted, and removed
+   * from this set before scheduling the next retry.
+   */
+  private _pendingOps: Set<Promise<void>> = new Set();
+
+  /**
+   * Whether this BulkWriter instance has started to close. Afterwards, no
+   * new operations can be enqueued, except for retry operations scheduled by
+   * the error handler.
+   */
+  private _closing = false;
 
   /**
    * Rate limiter used to throttle requests as per the 500/50/5 rule.
    */
-  private rateLimiter: RateLimiter;
+  private readonly _rateLimiter: RateLimiter;
+
+  /**
+   * The user-provided callback to be run every time a BulkWriter operation
+   * successfully completes.
+   */
+  private _successFn: (
+    document: firestore.DocumentReference,
+    result: WriteResult
+  ) => void = () => {};
+
+  /**
+   * The user-provided callback to be run every time a BulkWriter operation
+   * fails.
+   */
+  private _errorFn: (error: BulkWriterError) => boolean = error => {
+    const retryCodes = getRetryCodes('batchWrite');
+    const shouldRetry =
+      error.code !== undefined &&
+      retryCodes.includes(error.code) &&
+      error.failedAttempts < MAX_RETRY_ATTEMPTS;
+    logger(
```

Comment:
As proposed in API meeting, we should move this logging to the callsite. This will also allow to us to log user-customized retries.